blinding fixes

instagibbs opened the pull request:
edit: PR has moved on to change more behavior, please read the entire thread to get the gist
----------------
Current behavior leaves much to be desired: https://github.com/ElementsProject/elements/issues/137

This set of changes:

https://github.com/ElementsProject/elements/commit/b50da101051dc9a362843578ee2c07b1ec61e7c7 : Disables changePosition argument functionality. It was only used in `fundrawtransaction` and requires more thought on how to handle the user requesting multiple change outputs at specific places. I'm honestly not sure of the use-case to begin with, so removing it for now seems pertinent.

https://github.com/ElementsProject/elements/commit/d9905f06206826ab97e2e011c9d024af06635e36 Help text fixup

https://github.com/ElementsProject/elements/commit/318e01d53df15eee76468c98390cbf3167206be2 : is the big one. We need to make sure that any FundTransaction-based call to CreateTransaction recovers the original unblinded output set. This means the transaction no longer receives blinded change, or dummy OP_RETURN outputs to blind in the case of a single blinding factor. A regression of sorts would be that the user may have to manually append a blinded OP_RETURN output to the transaction after calling `fundraw` and before `blindraw`. Another side-effect is that the user can now actually call `fundraw` repeatedly on the same transaction, as previously it would encounter a blinded output and not know what to do.

https://github.com/ElementsProject/elements/commit/86f325254e077a3470a89e41646a22e59226147d : This fixes the issue of the user repeatedly funding a transaction, and the previous fees compounding on top of new fee estimates/additions. Just drop all previously-attached fees and let it work.

This resolves https://github.com/ElementsProject/elements/issues/137

jonasnick commented on src/wallet/wallet.cpp:
```diff
@@ -2404,18 +2408,15 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, bool ov
-        tx.wit.vtxoutwit.resize(tx.vout.size()-1);
-        tx.wit.vtxoutwit.insert(tx.wit.vtxoutwit.begin() + nChangePosInOut,  wtx.wit.vtxoutwit[nChangePosInOut]);
+    // Wipe outputs and output witness and re-add one by one
+    wtx.wit.vtxoutwit.resize(wtx.vout.size());
```
Why is vtxoutiwit resized?

instagibbs replied:
we're iterating over the size of the vout, and I just want to make sure a witness actually exists, blank or not for each.

jonasnick commented on src/wallet/wallet.cpp:
```diff
@@ -2529,6 +2537,7 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
                 BOOST_FOREACH (const CRecipient& recipient, vecSend)
                 {
                     CTxOut txout(recipient.asset, recipient.nAmount, recipient.scriptPubKey);
+                    txout.nNonce.vchCommitment = std::vector<unsigned char>(recipient.confidentiality_key.begin(), recipient.confidentiality_key.end());
```
What does that do?

instagibbs replied:
`createrawtransaction` sets the confidentiality keys of the outputs to be the CT address pubkey. Typically they stay there until `BlindTransaction`, but in the `fundrawtransaction` path you simply lose those keys since these newly made outputs are written back in FundTransaction.

instagibbs replied:
So all I'm doing here is writing them back to the newly-generated outputs.

instagibbs replied:
I think I have to do this for change outputs as well, otherwise we'll not actually blind the change output in a follow-up `blindtransaction` call.

jonasnick commented on src/wallet/wallet.cpp:
```diff
@@ -2878,26 +2877,25 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
-                    assert(!output_assets[i].IsNull());
-                    wtxNew.SetBlindingData(i, vAmounts[i], output_pubkeys[i], output_blinds[i], output_assets[i], output_asset_blinds[i]);
+                if (sign) {
+                    assert(vAmounts.size() == output_pubkeys.size());
```
Why did you move the asserts?

instagibbs replied:
I think I over-thought this case. Probably doesn't have to be moved inside.

instagibbs commented:
fixed a nit, added the `OP_RETURN` output appending behavior to `(raw)blindraw` since it has been removed otherwise, and added a basic test for this behavior.

jonasnick commented:
I've been testing this PR with a script that randomly generates a tx with a random combination of confidential and unconfidential output, then funds and blinds it: https://gist.github.com/jonasnick/9a4d3bf0fa81e4fbf0eaa67176d0d72a

The current code doesn't work. In particular if there are confidential inputs, but no blinded outputs then it will throw "Transaction has no outputs to blind".

I've written a patch which makes the test script works: https://github.com/jonasnick/elements/commit/f96e19f1285e4330ad2fc6f903a2b430e512a885
It checks the conditions for adding an OP_RETURN output explicitly instead of trying to blind. Also it doesn't take confidential assets into account.
The downside compared to the try-and-add-output is that if the input blinding factors already cancel out then it would unnecessarily add an op_return.

instagibbs commented:
@jonasnick makes sense. I'll incorporate

instagibbs commented:
cherrypicked @jonasnick 's commit and fixed tests. Will squash before merging.

instagibbs commented:
Since we're moving towards the "don't throw an error unless it's something unexpected", I made the vacuous 0-blinds-in, 0-blinds-out case simply return the original hex.

jonasnick commented on src/rpc/rawtransaction.cpp:
```diff
@@ -804,14 +805,30 @@ UniValue rawblindrawtransaction(const UniValue& params, bool fHelp)
+    if (numPubKeys == 0) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, string("Transaction has no outputs to blind"));
+    }
+    if (BlindTransaction(input_blinds, input_asset_blinds, input_assets, input_amounts, output_value_blinds, output_asset_blinds, output_pubkeys, std::vector<CKey>(), std::vector<CKey>(), tx) != numPubKeys) {
```
This should be similar to what we're doing in blindraw: check for the conditions to add an OP_RETURN directly.

jonasnick commented on src/rpc/rawtransaction.cpp:
```diff
@@ -931,9 +951,22 @@ UniValue blindrawtransaction(const UniValue& params, bool fHelp)
+        CTxOut newTxOut(tx.vout.back().nAsset.GetAsset(), 0, CScript() << OP_RETURN);
+        tx.vout.push_back(newTxOut);
+        numPubKeys++;
+        output_pubkeys.push_back(pwalletMain->GetBlindingPubKey(newTxOut.scriptPubKey));
```
What exactly are the output_pubkeys?

instagibbs replied:
They're the public blinding keys for the recipient, in this case you.

jonasnick commented on src/wallet/wallet.cpp:
```diff
@@ -2828,30 +2840,63 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
+                int ret = BlindTransaction(input_blinds, input_asset_blinds, input_assets, input_amounts, output_blinds, output_asset_blinds,  output_pubkeys, vassetKeys, vtokenKeys, txNew);
+                assert(ret != -1);
+                if (ret != numToBlind) {
+                    strFailReason = _("Unable to blind the transactoin properly. This should not happen.");
```
transactoin typo

jonasnick commented on src/wallet/wallet.cpp:
```diff
@@ -2828,30 +2840,63 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
+                            output_blinds[onlyChangePos] = uint256();
+                            output_asset_blinds[onlyChangePos] = uint256();
+                        } else {
+                            strFailReason = _("Change output could not be blinded as their are no blinded inputs and no other blinded outputs.");
```
typo: their

jonasnick commented:
I'm very happy with this PR now. It passes the wallet test shell script I posted earlier. I think it would be more clear if `acceptvalidblindingfail` would be renamed to something like `ignoreblindingfail`. It's debatable if it should be true by default, but either way is ok for me. What's also nice with this PR is that now you can blindraw any transaction without checking before if it actually needs blinding.

Nits:
- it should be noted in the fundrawtransaction help text that existing fees will get dropped
- blindrawtransaction and rawblindrawtransaction need an updated helptext to clarify what exactly is happening in the various scenarios.
    - Additionally, in blindrawtransaction "If at least one of the inputs is confidential, at least one of the outputs must be." is not true anymore

Before a final ACK I'd like to look at a squashed version to verify that there are no remains of superseded commits.

instagibbs commented:
squashed and cleaned up history a bit, fixed a couple nits.

Changed to an rpc argument instead of a startup argument because users may want to bypass the default without restarting. By default we ignore blind fails.

jonasnick commented on src/wallet/rpcwallet.cpp:
```diff
@@ -488,7 +488,7 @@ UniValue sendtoaddress(const UniValue& params, bool fHelp)
 
     if (fHelp || params.size() < 2 || params.size() > 6)
```
s/6/7

jonasnick commented on src/rpc/rawtransaction.cpp:
```diff
@@ -822,11 +862,11 @@ UniValue blindrawtransaction(const UniValue& params, bool fHelp)
 {
     if (fHelp || (params.size() != 1 && params.size() != 2))
```
ignorefail would be another parameter (a 4th one)

instagibbs commented:
Fixed issues pointed out by @jonasnick and made it easier to use these options by moving up in ordering for `blindraw`.

jonasnick commented on src/rpc/rawtransaction.cpp:
```diff
@@ -820,21 +860,22 @@ UniValue rawblindrawtransaction(const UniValue& params, bool fHelp)
 UniValue blindrawtransaction(const UniValue& params, bool fHelp)
 {
-    if (fHelp || (params.size() != 1 && params.size() != 2))
+    if (fHelp || (params.size() < 1 && params.size() > 4))
```
should be params.size() < 1 || params.size() > 4

instagibbs commented:
fixed again...

jonasnick commented:
tACK a3f68d0db77bad1283d02bb8f822f36d09d9af2e
